include/grpc++/auth_context.h

```diff
@@ -34,15 +34,42 @@
 #ifndef GRPCXX_AUTH_CONTEXT_H
 #define GRPCXX_AUTH_CONTEXT_H
 
+#include <iterator>
 #include <vector>
 
 #include <grpc++/config.h>
 
+struct grpc_auth_context;
+struct grpc_auth_property;
+struct grpc_auth_property_iterator;
+
 namespace grpc {
+class SecureAuthContext;
 
 class AuthContext {
  public:
   typedef std::pair<grpc::string, grpc::string> Property;
+  class PropertyIterator
+      : public std::iterator<std::input_iterator_tag, const Property> {
+   public:
+    ~PropertyIterator();
+    PropertyIterator& operator++();
+    PropertyIterator operator++(int);
+    bool operator==(const PropertyIterator& rhs) const;
+    bool operator!=(const PropertyIterator& rhs) const;
+    const Property operator*();
+
+   private:
+    friend SecureAuthContext;
+    PropertyIterator();
+    PropertyIterator(const grpc_auth_property* property,
+                     const grpc_auth_property_iterator* iter);
+    const grpc_auth_property* property_;
+    // The following items form a grpc_auth_property_iterator.
```

Comment:
It is unfortunate that you have to do that but I don't really see a way around it (except having a pointer to a `grpc_auth_property_iterator` but that would mean an extra alloc.